### PR TITLE
GET API for /datasets/inventory/ to access dataset artifacts

### PR DESCRIPTION
**Pull request by siddardh-ra**
Define Pbench server API to access dataset artifacts (including tarball)

PBENCH-625

Add a new `/datasets/inventory/` API to return the files as byte stream for the specified `dataset` with the `path` as URI parameters.

### CODE REVIEW

File: lib/pbench/server/filetree.py
Comment (dbutenhof) [inline lib/pbench/server/filetree.py:923]: This will find *controllers* rather than *datasets*; e.g., `dhcp31-194.perf.lab.eng.bos.redhat.com` rather than `pbench-user-benchmark_loop.3_2022.06.14T20.38.11`

Also, while right now `dataset.name` will match the tarball name, and I'll need to change some aspects of `FileTree` anyway when I change that, I'd rather not build in another instance that needs to be fixed later.

My idea had been to use `file_tree.find_dataset(dataset.name)` instead. If the dataset exists in the file tree, you'll get back a `Tarball`, and `tarball.unpacked` is the `Path` to the unpacked inventory if it exists. Then you can just search from that `Path` for the file you want.

**Review by dbutenhof — CHANGES_REQUESTED** [review]
Great start. I haven't reviewed the unit tests yet. I'll get through that later, but I wanted to post the comments I've been collecting all day...

File: lib/pbench/server/filetree.py
Comment (webbnh) [inline lib/pbench/server/filetree.py:926]: Is this the appropriate exception to raise?  (We're raising this because a file system artifact is missing...if the _dataset_ were missing, we never would have gotten this far, right?)

File: lib/pbench/server/api/resources/datasets_inventory.py
Comment (webbnh) [inline lib/pbench/server/api/resources/datasets_inventory.py:80]: The `except` block here is intended to recover from an error in `find_inventory()`, right?  If that's the case, then you should move everything after it out of the `try` block.

File: lib/pbench/server/api/resources/datasets_inventory.py
Comment (webbnh) [reply lib/pbench/server/api/resources/datasets_inventory.py:77]: As I think of it, I don't think that the `exists()` test will ever fail:  if the file were not found, then `find_inventory()` would raise an exception and control would never reach here.  So, if `file_path` is not `is_file()`, then it _is present_ but it's not anything that we're willing to deal with (so, we should abort, but not with `NOT_FOUND` and not claiming that the file isn't present).

File: lib/pbench/server/api/__init__.py
Comment (webbnh) [reply lib/pbench/server/api/__init__.py:95]: In this case, `path` is always relative to the "root of the dataset", so, if the leading slash isn't present it can and should be assumed.  (But, yeah, I would like to avoid the `//` situations.)

> So I think what we want here is f"{base_uri}/datasets/inventory/<string:dataset><path:path>" ... assuming that Flask can handle that, at least, which isn't entirely clear from the documentation.

I don't see how it could handle that...how would it know where `dataset` ends and `path` starts?  I think that URI parameters have to be separated by slashes...which isn't a problem in this case because we can either assume the leading slash on `path` or glue it in there during processing.

File: lib/pbench/server/api/resources/datasets_inventory.py
Comment (webbnh) [reply lib/pbench/server/api/resources/datasets_inventory.py:28]: I second the motion to remove "unpacked".  🙂

File: lib/pbench/server/filetree.py
Comment (webbnh) [reply lib/pbench/server/filetree.py:925]: This will be moot if you follow Dave's `find_dataset()` suggestion above.

File: lib/pbench/server/filetree.py
Comment (webbnh) [reply lib/pbench/server/filetree.py:926]: This, too, will be moot if you follow Dave's `find_dataset()` suggestion above.

File: lib/pbench/server/api/__init__.py
Comment (dbutenhof) [reply lib/pbench/server/api/__init__.py:95]: Yeah, you're likely right; in that case I believe there's Flask voodoo to map a missing parameter to a default instead of requiring the `//` for root, but it seems a bit ugly.

Right now on Siddardh's server, for example, `<dataset>/` returns 404 "The requested URL was not found on the server." (However, so does `//`, presumably meaning that the doubled slash isn't even enough to get Flask to process a `path` type parameter regardless of our aesthetic concerns!)

File: lib/pbench/server/filetree.py
Comment (dbutenhof) [reply lib/pbench/server/filetree.py:923]: I was confused here; since your API actually works on your test server, I had to come back and look again. And, yeah, you're using `rglob`, which is recursive, so it will iterate through *the full file tree of every unpacked tarball in every controller* under the `incoming` base directory until it finds a `.name` matching the dataset. (And although an inappropriate match is probably unlikely, this code doesn't even care whether the match is a directory, and it probably should.) Regardless, this could take a long time and waste a lot of resources.

There are better ways to code this directly; however, again, using `.find_dataset` to locate the named dataset and looking at the `.unpacked` attribute on the object instance would be a more consistent solution.

File: lib/pbench/test/unit/server/test_datasets_inventory.py
Comment (dbutenhof) [reply lib/pbench/test/unit/server/test_datasets_inventory.py:50]: Seems this was copied from `test_datasets_metadata.ps`, which had to deal with some special cases in both `PUT` and `GET` testing... which doesn't apply here.

In a previous comment, Webb commented about trying to reuse these query wrappers. I've thought about trying to refactor the unit test hierarchy again to pull out another level on top of (or rather underneath) what Nikhil factored out into the `Common` class. Never got around to doing it, though. (And honestly if I had that time and motivation I'd rather turn the API tests into true unit tests and true functional tests rather than preserve this mocked request half-state.)

Comment (webbnh) [reply lib/pbench/test/unit/server/test_datasets_inventory.py:50]: So...that's a "yes"?  😆

**Comment by siddardh-ra** [comment]
> This is an excellent initial cut. I found a few typos, comment tweaks, code suggestions, etc., but there are some more substantial issues to address (like missing parameters).

Hi @webbnh, 
In the recent commit, I addressed most of the enhancements and corrections as you suggested :)

> 
> Some of the test code looks like it is copies of other test code. We need to resist this to the fullest extent possible. If we need the code in two places, we should arrange to share it. (And, enabling refactoring like that would best be done as a separate PR which this PR can be built on.)

Got it, And yeah, I will refactor the common code for the test case in the separate PR

File: lib/pbench/server/api/resources/datasets_inventory.py
Comment (siddardh-ra) [reply lib/pbench/server/api/resources/datasets_inventory.py]: I used this function mainly due to an issue I faced during mocking send_file. But I got a suggestion from Dave regarding this, and in a recent commit, I removed this and directly mocked send_file.

**Comment by webbnh** [comment]
> I will refactor the common code for the test case in the separate PR

Do you intend to have that PR merged before or after this one?  (Should I hold off on re-reviewing this PR until that PR is up, because you're going to make this PR use the refactored code provided by that PR?  Yes?  Please?  😂)

File: lib/pbench/server/api/resources/datasets_inventory.py
Comment (npalaska) [reply lib/pbench/server/api/resources/datasets_inventory.py:64]: We don't use the ad-hoc unpack yet but this question came to my mind what should we do when the dataset is still not unpacked, do we want to explicitly unpack the dataset when the owner wants to see files within the dataset or should we let the owner do unpack first and we fail here?

Comment (dbutenhof) [reply lib/pbench/server/api/resources/datasets_inventory.py:64]: Before we release this, we'll have S3 and a cache manager, so I don't see any point in worrying about unpacking here. There's a small window between `PUT` and running the unpack tool where we might not have an `INCOMING` directory; we shouldn't blow things up by dereferencing a `None`, but I'd rather not waste any effort beyond that.

File: lib/pbench/test/unit/server/test_datasets_inventory.py
Comment (npalaska) [reply lib/pbench/test/unit/server/test_datasets_inventory.py:68]: Also, how is this working, to get a token with a user "user" we need to first register the "user" and where are we registering this user? Is this leaking user registration from the `TestUserAuthentication` class of `test_authorization`?

File: lib/pbench/test/unit/server/test_datasets_inventory.py
Comment (dbutenhof) [reply lib/pbench/test/unit/server/test_datasets_inventory.py:55]: Webb brought this up earlier; it's because Siddardh carried this code forward from the metadata tests, which faced some special requirements due to multiple API calls (trying to leverage the same general helpers) within a single test case. Because the second call ended up trying to log in again and got the duplicate `auth_token` issue, it was "cleaner" (or at least expedient) to log out. Not necessary here or in most other tests; and hopefully won't be necessary at all once we go to Keycloak.

File: lib/pbench/test/unit/server/test_datasets_inventory.py
Comment (dbutenhof) [reply lib/pbench/test/unit/server/test_datasets_inventory.py:68]: Fixtures. Those magical little Pytest fairy dust generators ... 😆 

These tests pull `provide_metadata` (which this test class doesn't actually need, but again was inherited from the metadata test, which did need it). And `provide_metadata` requires datasets, which require registered users, both `drb` and `test`. So the various forms of `query_x_as` fixtures here, which all depend on `provide_metadata`, pull all that in.

File: lib/pbench/test/unit/server/test_endpoint_configure.py
Comment (dbutenhof) [inline lib/pbench/test/unit/server/test_endpoint_configure.py:55]: Siddardh; you need to rebase on `main` again, and you'll need to update this because the trailing `_` in the key and the trailing url `/` characters are gone now. As written, this test will fail once you merge/rebase.

File: lib/pbench/test/unit/server/test_datasets_inventory.py
Comment (dbutenhof) [inline lib/pbench/test/unit/server/test_datasets_inventory.py:11]: Is there a reason for conditionalizing these types?

And as I saw this, I also realize that I hadn't even consciously noticed you were importing all of `typing` and using `t.Union` and such throughout. That wastes a bit more space, and it's not the way we do type hints in every other file in the repo, which makes it look a bit odd... and potentially makes the file just a bit harder to read.

I'd prefer you do this like everywhere else, and just `from typing import Callable, Union, ...`

File: lib/pbench/test/unit/server/test_datasets_inventory.py
Comment (npalaska) [reply lib/pbench/test/unit/server/test_datasets_inventory.py:68]: Also, you can import the generic password from `lib/pbench/test/unit/server/conftest.py` instead of hardcoding it here

File: lib/pbench/test/unit/server/test_datasets_inventory.py
Comment (dbutenhof) [reply lib/pbench/test/unit/server/test_datasets_inventory.py:25]: Yeah, a long time ago Nikhil factored out some common Elasticsearch tests into `common.py` to be shared. But we have both specialized Elasticsearch APIs and non-Elasticsearch APIs (like this one) that don't really fit the mold.

At some point, we need to step back and factor out another common test class (or two) to share more stuff cleanly; it just hasn't gotten to be anyone's priority yet.

File: lib/pbench/server/api/resources/datasets_inventory.py
Comment (webbnh) [reply lib/pbench/server/api/resources/datasets_inventory.py:69]: > The second argument shouldn't be a Path, just a string.

The second argument to _what_?  (Nikhil seems to understand what you mean, @dbutenhof, but I'm not sure I'm following...I hope Siddardh is doing better than me.  😆)

Is `dataset_location` a `Path`??  (Gosh I miss having explicit types....)

If so, then I concur that it would be best not to convert `path` into a `Path`, as the `pathlib` rules for joining absolute paths can produce some counterintuitive results in certain situations (of which this might be one!).

**Comment by webbnh** [comment]
> I will refactor the common code for the test case in the separate PR

How is this going, Siddardh?

You're going to tell me when I should re-review this PR, right?  (😁)

File: lib/pbench/server/api/resources/datasets_inventory.py
Comment (siddardh-ra) [reply lib/pbench/server/api/resources/datasets_inventory.py:69]: >Is dataset_location a Path??

Yeah @webbnh you're right :+1: , `dataset_location` will return a `Path` and the `path` variable is a string. In order to append these two, I'm converting the latter into `Path`[`pathlib`] for further references

Comment (webbnh) [reply lib/pbench/server/api/resources/datasets_inventory.py:69]: Right, @siddardh-ra, but you don't need to:  the [`/` operator when applied to a `Path`](https://docs.python.org/3/library/pathlib.html#operators) will accept a `str` for its other argument.  So, converting it to a `Path` first is likely a waste of time and space.

**Review by riya-17 — DISMISSED** [review]
Looks Good

**Review by webbnh — DISMISSED** [review]
Looks very nice.

I found a couple of nits which you should probably fix and a couple of others for you to consider.

**Review by riya-17 — APPROVED** [review]

**Pull request merged**
